src/AccelStepperFirmata.cpp

```diff
@@ -0,0 +1,430 @@
+        
+      }
+
+      else if (stepCommand == MULTISTEPPER_STOP) {
+
+        groupIsRunning[deviceNum] = false;
+        
+      }
+    }
+    return true;
+    
+  }
+  return false;
+}
+
+/*==============================================================================
+ * SETUP()
+ *============================================================================*/
+
+void AccelStepperFirmata::reset()
+{
+  for (byte i = 0; i < MAX_STEPPERS; i++) {
+    if (stepper[i]) {
+      free(stepper[i]);
+      stepper[i] = 0;
+    }
+  }
+  numSteppers = 0;
+  
+  for (byte i = 0; i < MAX_GROUPS; i++) {
+    if (group[i]) {
+      free(group[i]);
+      group[i] = 0;
+    }
+  }
+  numGroups = 0;
+}
+
+/*==============================================================================
+ * Helpers
+ *============================================================================*/
+
+float AccelStepperFirmata::decodeCustomFloat(byte arg1, byte arg2, byte arg3, byte arg4)
+{
+  long l4 = (long)arg4;
+  long significand = (long)arg1 | (long)arg2 << 7 | (long)arg3 << 14 | (l4 & 0x03) << 21;
+  float exponent = (float)(((l4 >> 2) & 0x0f) - 11);
+  bool sign = (bool)((l4 >> 6) & 0x01);
+  float result = (float)significand;
+
+  if (sign) {
+    result *= -1;
+  }
+
+  result = result * powf(10.0, exponent);
+
+  return result;
+}
+
+long AccelStepperFirmata::decode32BitSignedInteger(byte arg1, byte arg2, byte arg3, byte arg4, byte arg5)
+{
+  long result = (long)arg1 | (long)arg2 << 7 | (long)arg3 << 14 | (long)arg4 << 21 | (((long)arg5 << 28) & 0x07);
+
+  if ((long)arg5 >> 3 == 0x01) {
+    result = result * -1;
+  }
+
+  return result;
+}
+
+void AccelStepperFirmata::encode32BitSignedInteger(long value, byte pdata[])
+{
+  bool inv = false;
+
+  if (value < 0) {
+    inv = true;
+    value = value * -1;
+  }
+  
+  pdata[0] = value & 0x7f;
+  pdata[1] = (value >> 7) & 0x7f;
+  pdata[2] = (value >> 14) & 0x7f;
+  pdata[3] = (value >> 21) & 0x7f;
+  pdata[4] = (value >> 28) & 0x7f;
+  
+  if (inv == true) {
+    pdata[4] = pdata[4] | 0x08;
+  }
+}
+
+/*==============================================================================
+ * LOOP()
+ *============================================================================*/
+void AccelStepperFirmata::update()
+{
+  bool stepsLeft;
+
+  if (numGroups > 0) {
+    // if one or more groups exist,  update their position
+    for (byte i = 0; i < MAX_GROUPS; i++) {
+      if (group[i] && groupIsRunning[i] == true) {
+        stepsLeft = group[i]->run();
+        
+        // send command to client application when stepping is complete
+        if (stepsLeft != true) {
+          groupIsRunning[i] = false;
+          reportGroupComplete(i);
+        }
+        
+      }
+    }
+  }
+  
+  if (numSteppers > 0) {
+    // if one or more stepper motors are used, update their position
+    for (byte i = 0; i < MAX_STEPPERS; i++) {
+      if (stepper[i] && isRunning[i] == true) {
+        stepsLeft = stepper[i]->run();
```

Comment:
If the user calls `stop`, `stepsLeft` will be false even if the step sequence is not complete because the `run()` function returns false if `_speed == 0` which is the case if `stop` is called. This results in the `done` event being triggered when the STEPPER_STOP message is sent.

One way to get around this is to change the condition below to the following:
```c++
if (!stepper[i]->isRunning()) {
  ...
}
```
the `isRunning()` method checks if the speed == 0 AND the step sequence is complete.

A change like this would require rethinking the callback strategy in firmata.js since you could be left with un unhandled event in the case that stop is called and the event is listening for 'done' and the sequence is never resumed.

Reply:
This however depends on whether or not a user should be able to resume after sending a `STEPPER_STOP` message.

Reply:
After further review, I've determined no change is necessary here. I was unaware initially that the stop() behavior was terminal.